File: solutions/haskell/143-coin-change.hs
{-
  Problem 143: Coin Change (LeetCode 322)
  Difficulty: Med
  Language: Haskell
-}
coinChange :: [Int] -> Int -> Int
coinChange coins amount =
  let inf = amount + 1
      dp = [if i == 0 then 0 else inf | i <- [0..amount]]
      update table coin =
        [if i >= coin then min (table !! i) (table !! (i - coin) + 1)
         else table !! i | i <- [0..amount]]
      result = foldl update dp coins
  in if result !! amount >= inf then -1 else result !! amount

File: solutions/rust/143-coin-change.rs
/**
 * Problem 143: Coin Change (LeetCode 322)
 * Difficulty: Med
 * Language: Rust
 */
fn coin_change(coins: &[i32], amount: usize) -> i32 {
    let inf = amount + 1;
    let dp: Vec<usize> = (0..=amount).map(|i| if i == 0 { 0 } else { inf }).collect();
    let result = coins.iter().fold(dp, |table, &coin| {
        (0..=amount).map(|i| {
            if i >= coin as usize { table[i].min(table[i - coin as usize] + 1) }
            else { table[i] }
        }).collect()
    });
    if result[amount] >= inf { -1 } else { result[amount] as i32 }
}
